File: Front/game_src/game_offline/Ball.js
import { Config } from './Config.js'
import { Countdown } from './Countdown.js';
import { PowerUp } from './PowerUp.js'

export class Ball {
	constructor( scene, color, powerUpsEnabled) {
		this.scene = scene;
		this.arena = this.scene.arena;
		this.score = this.scene.score;
		this.color = color;

		// Ball propreties
		this.radius = 0.03;
		this.velocityX = 0.015;
		this.velocityY = 0.015;
		this.maxSpeed = 0.03

		//Ball Drawing
		this.geometry = new Config.THREE.SphereGeometry(this.radius)
		this.material = new Config.THREE.MeshLambertMaterial({color: this.color})
		this.mesh = new Config.THREE.Mesh(this.geometry, this.material);
		this.mesh.position.set(0, 0, this.radius + this.arena.arenaDepth / 2);

		//Countdown
		this.countdown = new Countdown(this.scene, this.arena);

		//PowerUps
		this.powerUpsEnabled = powerUpsEnabled;
		this.consecutiveGoals = { left: 0, right:0 };
		this.powerUp = null;
	}

	start( direction ) {
		const angle = (Math.random() * Math.PI / 4) - Math.PI / 8; // -22.5 to 22.5 degrees
		const speed = 0.025;
		this.velocityX = speed * Math.cos(angle) * direction;
		this.velocityY = speed * 0;
	}

	reset() {
		this.mesh.position.set(0, 0, this.radius + this.arena.arenaDepth / 2);
	
		// Initial velocity
		this.velocityX = 0;
		this.velocityY = 0;
	}

	update( leftPaddle, rightPaddle ) {
		// Check for wall collisions
		this.checkWallCollision(leftPaddle, rightPaddle);
	
		// Check for paddle collisions
		this.checkPaddleCollision(leftPaddle);
		this.checkPaddleCollision(rightPaddle);

		if (this.powerUp){
			if (!this.powerUp.isActive && this.powerUp.checkExpired()) {
				
				this.powerUp.remove();
				this.powerUp = null;
				return ;
			}
			if (!this.powerUp.isActive && this.powerUp.checkCollision()){
				
				this.powerUp.applyPowerUp();
				this.powerUp.remove();
			}
		}
	}

	removePowerUp( goalOwner ) {
		if (this.powerUp && this.powerUp.owner.side !== goalOwner.side){
			this.powerUp.remove();
			if (this.powerUp.isActive){
				this.powerUp.removeAbility();
			}
			this.powerUp = null;
		}
		else {
			if (this.powerUp && this.powerUp.owner.clan === "VERTEX"){
				if (this.powerUp.isActive){
					this.powerUp.removeAbility();
				}
			}
		}
	}

	checkWallCollision ( leftPaddle, rightPaddle ) {
		const prevBallY = this.mesh.position.y;

		// Move ball
		this.mesh.position.x += this.velocityX;
		this.mesh.position.y += this.velocityY;

		// Bounce off top and bottom with slight speed increase
		if (this.mesh.position.y + this.radius > this.arena.arenaHeight / 2 || 
		    this.mesh.position.y - this.radius < -this.arena.arenaHeight / 2) {
		  this.velocityY = -this.velocityY * 1.025;
		  this.mesh.position.y = prevBallY; // Prevent sticking to walls
		}

		// left scored
		if (this.mesh.position.x + this.radius > this.arena.arenaWidth / 2) {
			leftPaddle.score++;
			if (this.powerUpsEnabled){
			this.removePowerUp(leftPaddle);
				this.consecutiveGoals.right = 0;
				this.consecutiveGoals.left++;
				if (this.consecutiveGoals.left === 3){
					this.consecutiveGoals.left = 0;
					if (this.powerUp){
						this.powerUp.remove();
						this.powerUp = null;
					}
					this.powerUp = new PowerUp(this.scene, leftPaddle);
					this.powerUp.spawn();
				}
			}

			this.score.updateScore(leftPaddle.score, rightPaddle.score);

			if (leftPaddle.score === this.scene.goalLimit){
				this.scene.gameOverSequence();
				return ;
			}
			this.countdown.startCountdown("r");
			leftPaddle.reset();
			rightPaddle.reset();
			this.reset();
			this.start(1);
		}

		// right scored
		else if (this.mesh.position.x - this.radius < -this.arena.arenaWidth / 2) {
			rightPaddle.score++;
			if (this.powerUpsEnabled){
			this.removePowerUp(rightPaddle);
				this.consecutiveGoals.left = 0;
				this.consecutiveGoals.right++;
				if (this.consecutiveGoals.right === 3){
					this.consecutiveGoals.right = 0;
					if (this.powerUp){
						this.powerUp.remove();
						this.powerUp = null;
					}
					this.powerUp = new PowerUp(this.scene, rightPaddle);
					this.powerUp.spawn();
				}
			}
			this.score.updateScore(leftPaddle.score, rightPaddle.score);

			if (rightPaddle.score === this.scene.goalLimit){
				this.scene.gameOverSequence();
				return ;
			}
			this.countdown.startCountdown("l");
			leftPaddle.reset();
			rightPaddle.reset();
			this.reset();
			this.start(-1);
		}

		// Ensure ball doesn't exceed max speed
		const currentSpeed = Math.sqrt(this.velocityX ** 2 + this.velocityY ** 2);
		if (currentSpeed > this.maxSpeed) {
			const scale = this.maxSpeed / currentSpeed;
			this.velocityX *= scale;
			this.velocityY *= scale;
		}
	}

	checkPaddleCollision( paddle ) {
		const paddleX = paddle.mesh.position.x;
		const paddleY = paddle.mesh.position.y;
		const ballX = this.mesh.position.x;
		const ballY = this.mesh.position.y;

		// Basic collision detection
		if (ballX - this.radius < paddleX + paddle.paddleWidth / 2 &&
		    ballX + this.radius > paddleX - paddle.paddleWidth / 2 &&
		    ballY - this.radius < paddleY + paddle.paddleHeight / 2 &&
		    ballY + this.radius > paddleY - paddle.paddleHeight / 2) {
      
			// Calculate relative impact position (-1 to 1, from paddle bottom to top)
			const relativeIntersectY = (paddleY - ballY) / (paddle.paddleHeight / 2);
      
			// Calculate new angle (-60 to 60 degrees)
			const bounceAngle = relativeIntersectY * Math.PI / 3;
      
			// Calculate paddle's momentum
			const paddleMomentum = paddle.lastY - paddleY;
      
			// Determine if this is the left or right paddle
			const isLeftPaddle = paddleX < 0;
      
			// Calculate new velocities
			const speed = Math.sqrt(this.velocityX ** 2 + this.velocityY ** 2) * 1.05; // Slight speed increase
			this.velocityX = speed * Math.cos(bounceAngle) * (isLeftPaddle ? 1 : -1);
			this.velocityY = -speed * Math.sin(bounceAngle) - paddleMomentum * 0.2; // Add paddle momentum
      
			// Move ball out of paddle to prevent multiple collisions
			this.mesh.position.x = paddleX + (isLeftPaddle ? 1 : -1) * 
			(paddle.paddleWidth / 2 + this.radius);
		}
	}
}

File: Back/backend/game/ball.py
import random
import threading
import math
from .power_up import PowerUp

class Ball:
    """
    Represents the ball in the Pong game with movement and collision logic.
    """
    def __init__(self, arena):
        """
        Initialize the ball with default properties.
        
        :param arena: Arena instance the ball belongs to
        """
        self.arena = arena
        
        # Ball properties
        self.radius = 0.03
        self.max_speed = 0.03

        # Power Up
        self.consecutive_goals = {'left': 0, 'right': 0}
        self.power_up = None
        self.is_power_up_allowed = None
        
        # Initial position and velocity
        self.reset()

    def reset(self, direction=None):
        """
        Reset ball to initial position and velocity.
        
        :param direction: Initial ball direction (-1 or 1)
        """
        # Center of the arena
        self.x = 0
        self.y = 0
        
        # Initial velocity
        if direction is None:
            direction = random.choice([-1, 1])
        
        # Random angle within -22.5 to 22.5 degrees
        angle = (random.random() * math.pi / 4) - math.pi / 8
        speed = 0.025
        
        self.vx = speed * math.cos(angle) * direction
        self.vy = 0

    def update(self, paddles):
        """
        Update ball position based on current velocity.
        """
        # Move ball
        self.x += self.vx
        self.y += self.vy

        self.prev_y = self.y
        # Bounce off top and bottom walls
        if (self.y + self.radius > self.arena.arena_height / 2 or 
            self.y - self.radius < -self.arena.arena_height / 2):
            self.vy = -self.vy * 1.025
            self.y = self.prev_y

        # Ensure ball doesn't exceed max speed
        current_speed = math.sqrt(self.vx ** 2 + self.vy ** 2)
        if current_speed > self.max_speed:
            scale = self.max_speed / current_speed
            self.vx *= scale
            self.vy *= scale

        if self.power_up:
            if not self.power_up.is_active and self.power_up.check_expired():
                self.power_up = None
                return
            
            if not self.power_up.is_active and self.power_up.check_collision():
                self.power_up.apply_power_up()
                threading.Timer(self.power_up.activation_duration, self.clear_power_up).start()

    def clear_power_up(self):
        self.power_up = None

    def bounce_off_paddle(self, paddle, is_left_paddle):
        """
        Calculate ball bounce when hitting a paddle.
        
        :param paddle: Paddle instance
        :param is_left_paddle: Boolean indicating paddle side
        """
        # Calculate bounce angle
        bounce_angle = paddle.calculate_bounce_angle(self.y)
        
        # Calculate current speed
        speed = math.sqrt(self.vx ** 2 + self.vy ** 2) * 1.05
        
        # Update velocities
        self.vx = speed * math.cos(bounce_angle) * (1 if is_left_paddle else -1)
        self.vy = -speed * math.sin(bounce_angle)

        # Move ball out of paddle to prevent multiple collisions
        paddle_x = paddle.paddle_x
        self.x = paddle_x + (1 if is_left_paddle else -1) * (paddle.paddle_width / 2 + self.radius)

    def check_goal(self, paddles):
        """
        Check if a goal has been scored.
        
        :param paddles: Dict of paddles
        :return: Side that scored (None if no goal)
        """
        goal_side = None
        if self.x + self.radius > self.arena.arena_width / 2:
            goal_side = 'left'
            paddles['left'].score += 1

            if (self.is_power_up_allowed):
                paddles['left'].consecutive_goals += 1
                paddles['right'].consecutive_goals = 0

                if paddles['left'].consecutive_goals == 3:
                    paddles['left'].consecutive_goals = 0
                    self.power_up = PowerUp(self.arena, paddles['left'], self)
                    self.power_up.spawn()

        elif self.x - self.radius < -self.arena.arena_width / 2:
            goal_side = 'right'
            paddles['right'].score += 1

            if (self.is_power_up_allowed):
                paddles['right'].consecutive_goals += 1
                paddles['left'].consecutive_goals = 0

                if paddles['right'].consecutive_goals == 3:
                    paddles['right'].consecutive_goals = 0
                    self.power_up = PowerUp(self.arena, paddles['right'], self)
                    self.power_up.spawn()

        return goal_side
